### PR TITLE
fix(api): harden dashboard mutation routes

### LINKED ISSUE #1458: chore(ui): dashboard polish follow-ups from review
## Summary

Deferred polish for the dashboards feature, batched for one cleanup pass once the epic's main tasks land — none of these block shipping, all are worth closing before the feature is called done-done.

## Tasks

- [ ] **`DashboardGrid` mount-fire comment vs behavior** — the layout-init comment describes a PATCH the lazy-init makes impossible; align the comment with the actual never-write-on-view behavior (or change the init if persisting library compaction is desired).
- [ ] **`DashboardGrid` effect deps** — the unmount-flush effect closes over `onLayoutChange` with `[]` deps (lint warning); add the dep or hold the callback in a ref.
- [ ] **`DetectorWidget` loading state** — loading and failed findings both render "—"; distinguish them and drop the unused loading flag (lint warning).
- [ ] **Table renderer formatting** — format only value columns: a string dimension that parses as a number (e.g. a name like "2024") must not render as "2,024".
- [ ] **Time normalization consistency** — pass widget query `start_time`/`end_time` through `to_utc_naive` like every other ClickHouse call site (harmless today for ISO payloads; protects direct API callers sending naive local times).
- [ ] **Rate limiting** — the widgets endpoints have no limiter while the analogous traces meta endpoints do; add the shared read-bucket limiter for consistency (the distinct-values cache already bounds repeat scan cost).
- [ ] **Name-lock escape hatch** — clearing the widget name never re-enables auto-naming; consider re-unlocking on empty. Decide on user feedback.
- [ ] **Auth-session boilerplate** — the `useAuthSession → sessionReady/user` block repeats across the dashboard data hooks; extract a small local helper.

## Done when

All boxes checked in one or two small PRs, with the lint warnings gone and a regression test for the table-formatting fix.

### PR DESCRIPTION
**Pull request by ka1kqi**
Eleventh PR in the stacked project-dashboards series (epic #1460), stacked on #1521. The hardening batch for the dashboard/widget CRUD routes, split out from the route-introduction PRs so it reviews as its own diff — this is the PR the "pre-hardening" notes in #1520/#1521 point at.

## What's here
- **Role gates** — all six mutating handlers now require the MEMBER role; a VIEWER-role workspace member could previously create, rewrite, and cascade-delete dashboards and widgets. Reads stay viewer-accessible. Reuses the existing `minRole` plumbing the api-key routes already use.
- **Layout validation** — the layout PATCH validates every entry as an `{i, x, y, w, h}` placement (string id, finite non-negative coordinates) and whitelists exactly those keys, so a single malformed entry can't be persisted into jsonb where it would crash the grid for every member, and properties like `static` can't be smuggled in.
- **Length caps** — widget titles cap at 100 chars and dashboard descriptions at 500, matching the existing 50-char name cap, so one member can't persist unbounded strings that every dashboard read ships to all members. (This is the fix for cubic's two findings on #1521.)
- **P2025 → 404** — a concurrent delete between the scoped `findFirst` and the write now maps to 404 instead of an unhandled 500, via a shared `isRecordGone` helper.
- ~130 lines of tests covering each hardening dimension; the dashboard rename-cap test keeps its exact 50-char boundary probe rather than reverting to the mistitled 101-char version.

closes #1458

---
## Summary by cubic
Hardened dashboard and widget mutation routes and shipped dashboards with tabs, a shared per‑project date filter, and a guided widget builder with live preview. Blocks viewer writes, validates layouts and strings, and makes dashboards safe to edit and easier to use; addresses Linear #1458.

- **New Features**
  - Dashboard UI: tabs, shared date filter with presets persisted across pages (URL param still shareable), and a drag/resize grid that saves; index redirects to the default dashboard; seeded Overview is read‑only; create dialog with a 50‑char name cap.
  - Widget builder: create/edit at `/widgets/new` and `/widgets/{id}/edit` with step‑by‑step spec, live preview using real queries/renderers over the shared range, and auto titles until the user edits.

- **Bug Fixes**
  - API hardening: require `Role.MEMBER` for POST/PATCH/DELETE via `requireProjectAuth(minRole)`; validate layout entries `{i,x,y,w,h}`; cap widget `title` (100) and dashboard `description` (500); map Prisma P2025 to 404 via `isRecordGone`.
  - UI fixes: redirect only on true “Dashboard not found”; hide edit controls until detail resolves; duplicate preserves `displayConfig`; create dialog can’t close while pending; builder remounts when widget changes; preview pauses for histogram‑blocked drafts; date filter restores pre‑paint to avoid flashing the default window.

<sup>Written for commit 4d307ed49d0db8e68fa0a95dfd26e3d558067f6d. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/traceroot-ai/traceroot/pull/1522?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: frontend/ui/src/app/api/projects/[projectId]/dashboards/[dashboardId]/route.ts [hunk 1]
```diff
@@ -43,11 +43,43 @@ export async function PATCH(req: NextRequest, { params }: RouteParams) {
...
+    const isPlacement = (v: unknown): boolean =>
+      typeof v === "object" &&
+      v !== null &&
+      !Array.isArray(v) &&
+      typeof (v as Record<string, unknown>).i === "string" &&
+      ((v as Record<string, unknown>).i as string).length <= 128 &&
+      ["x", "y", "w", "h"].every((k) => {
+        const n = (v as Record<string, unknown>)[k];
+        return typeof n === "number" && Number.isFinite(n) && n >= 0;
+      });
```
[the remark this answers is not in the document]
Comment (ka1kqi) [reply frontend/ui/src/app/api/projects/[projectId]/dashboards/[dashboardId]/route.ts:68]: The failure premise doesn't hold on our react-grid-layout version — we verified this empirically for the same finding on the widget-grid PR: RGL 2.x dropped the 1.x PropTypes validation, so an item smaller than its `minW`/`minH` renders at its stored size with zero errors (a 0×0 entry is just an invisible tile), can't be resized below the minimum, and self-heals to ≥2×2 on the first resize, which then persists the corrected layout. Reachability is also narrow: the grid and builder never emit `w`/`h` below the minimum — only a hand-crafted API PATCH can. That said, a size floor here is reasonable boundary hardening, and it's already tracked as an epic follow-up together with the read-side clamp discussed on the grid PR (clamping `w`/`h` up to the minimums when building the layout), so stale stored entries get fixed at render time too. Deferring to that follow-up rather than growing this slice.

**Review by XinweiHe — APPROVED** [review]
